src/core/Queue.cpp

```diff
@@ -193,87 +196,103 @@ bool Queue::isEmpty() const
   return m_queue.empty();
 }
 
-Queue::Command* Queue::update()
+void Queue::execute(Command *command)
 {
-  if (m_queue.empty())
+  // Find command in queue
+  auto it = std::find(m_queue.begin(), m_queue.end(), command);
+
+  // If this queue is not out of order, add previous event associated with
+  // previous command in queue as dependency
+  if (it != m_queue.begin() /* && Check if not out of order queue */)
   {
-    return NULL;
+    command->waitList.push_back((*std::prev(it))->event);
```

Comment:
I feel like this should be done at the point of enqueue, rather than at the point of execution, unless there's a good reason to do it here?

Reply:
Yes, as the code stands, it's more logical to do this at the point of enqueue. However, in the final out-of-order implementation (my next PR), there's the issue of flushing/finishing in out-of-order queues (see [here](https://github.com/fakenmc/Oclgrind/blob/ooo_queues/src/core/Queue.cpp#L199)). In this (to be) proposed implementation, when the queue is out-of-order, regular `clWaitForEvents()` calls don't flush the queue, but when calling `clFinish()`/`clFlush()` the queue gets flushed. Being flushed or not depends on the `flush` parameter. This code is at the point of execution since only at this stage we know if the queue is to be flushed or not.

Alternatively, we could place this code at the point of enqueue and add an extra method to the `Queue` class, something like `executeAll()`, which iterates over all commands in the queue and executes them, and is called by `clFinish()`. Since we'll be iterating over a queue which may have dependent commands removed during iteration, and we have to add a new method just for this case, I find the existing solution cleaner.

Let me know what you think.